File: LeetCode/Python/Special/kmp.py
def get_next(pattern: str) -> [int]:
    i = 0
    j = -1
    next = [0] * len(pattern)
    next[0] = -1

    while i < len(pattern) - 1:
        if j == -1 or pattern[i] == pattern[j]:
            i += 1
            j += 1
            next[i] = j
        else:
            j = next[j]
    return next


def kmp(text: str, pattern: str) -> int:
    next = get_next(pattern)
    i = j = 0

    while i < len(text) and j < len(pattern):
        if j == -1 or text[i] == pattern[j]:
            i += 1
            j += 1
        else:
            j = next[j]
    if j == len(pattern):
        return (i - j)
    return -1


get_next("ABCDABD")
kmp("BBC ABCDAB ABCDABCDABDE", "ABCDABD")

File: NOwc0der/NOwc0der/String/KMP.swift
//
//  KMP.swift
//  NOwc0der
//
//  Created by Weslie on 3/31/19.
//  Copyright © 2019 weslie. All rights reserved.
//

private func getNext(pattern: [Character]) -> [Int] {
    var i = 0
    var j = -1
    var next = Array(repeating: 0, count: pattern.count)
    next[0] = -1
    
    while i < pattern.count - 1 {
        if j == -1 || pattern[i] == pattern[j] {
            i += 1
            j += 1
            if pattern[i] != pattern[j] {
                next[i] = j
            } else {
                next[i] = next[j]
            }
        } else {
            j = next[j]
        }
    }
    
    return next
}

func kmp(text: [Character], pattern: [Character]) -> Int {
    let next = getNext(pattern: pattern)
    
    var i = 0
    var j = 0
    
    while i < text.count && j < pattern.count {
        if j == -1 || text[i] == pattern[j] {
            i += 1
            j += 1
        } else {
            j = next[j]
        }
    }
    
    if j == pattern.count {
        return i - j
    }
    
    return -1
}

extension String {
    func kmpContains(_ anotherString: String) -> Bool {
        let str = Array(self)
        let pattern = Array(anotherString)
        return kmp(text: str, pattern: pattern) == -1 ? false : true
    }
}
